python/ray/tune/execution/ray_trial_executor.py

```diff
@@ -497,12 +533,90 @@ def _start_trial(self, trial: Trial) -> bool:
         self.restore(trial)
         self.set_status(trial, Trial.RUNNING)
 
-        self._staged_trials.discard(trial)
+        self._unstage_trial_with_resources(trial)
 
         if not trial.is_restoring:
             self._train(trial)
         return True
 
+    def _unstage_trial_with_resources(self, trial: Trial):
+        # Case 1: The trial we started was staged. Just remove it
+        if trial in self._staged_trials:
+            self._staged_trials.remove(trial)
+            return
+
+        # Case 2: We staged a trial "A" with the same resources, but our trial "B"
+        # was selected by the scheduler to run. The resource manager does not care
+        # about "trials", it just cares about resources being available. Thus we
+        # look for a staged trial with the same resource requirements and remove it
+
+        resource_request = trial.placement_group_factory
+        # Remove staged trial with same resource requirements
+        candidate_trial = None
+        for staged_trial in self._staged_trials:
+            staged_resources = staged_trial.placement_group_factory
+            if staged_resources == resource_request:
+                candidate_trial = staged_trial
+                break
+
+        if candidate_trial:
+            self._staged_trials.remove(candidate_trial)
```

Comment:
why do we keep track of `_staged_trials` in the first place, since everything is around resources? And seems we remove random trials that have the same resources request.

Reply:
This is existing logic. 

It's the callers responsibility to request resources and cancel resource requests. We only want to request each trial resources once, so we need to keep track for which trials we've already requested resources.

When we reuse actors we cancel the resource request from the trial that will reuse the cached actor. If that trial has not been staged (e.g. because max staging was reached and scheduler decided to start this trial), we can just remove another trial with the same resource requests instead.

Reply:
yeah understand the code has always been here. 

However, my understanding is that we shouldn't keep *which trials* are staged and has a bunch of comments explaining the concept of swapping them. 
Could we establish the minimal invariants (sort of tied to my one of my comments about high level invariants that this executor class is operating under). For example, we could state that at any given point, sum(staged_trial_number for any given resource request type) <= `executor._max_staged_actors` (is this correct?) Any many other constraints like this...

I feel like establishing this kind of invariants and state them clearly at the top of the class could greatly reduce the complexity of understanding how this class operates. Let me know if this makes sense.

Reply:
We are using the `staged_trials` variable in other places (to iterate), so for simplicity lets keep it here. It will likely be removed and definitely refactored in the actor manager execution refactor.

I've added much more extensive documentation for the invariants now - I hope this makes sense! ptal